active/0051-packaging.md

```diff
@@ -0,0 +1,338 @@
+const packagerFor = require('ember-cli/lib/broccoli/packager-for');
+
+module.exports = function(defaults) {
+  const app = new EmberApp(defaults, { });
+
+  const defaultPackager = packagerFor(app);
+  app.package = function(inputTree) {
+    // customise `inputTree`
+
+    return defaultPackager(inputTree);
+  }
+
+  return app.toTree();
+}
+```
+
+`packageFor` has the following signature:
+
+```typescript
+type Packager: (inputTree: BroccoliTree): BroccoliTree;
+
+function packagerFor(app: EmberApp): Packager;
+```
+
+`packagerFor` must return a function that returns a `BroccoliTree`.
+
+### Possible usages
+
+#### Debug/Analyse
+
+One of the applications of `package` API would be to run different analysis on the
+Ember applications. Take
+[broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser),
+for example. This could be easily incorporated into the build.
+
+```javascript
+// ember-cli-build.js
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+const packagerFor = require('ember-cli/lib/broccoli/packager-for');
+
+module.exports = function(defaults) {
+  const app = new EmberApp(defaults, { });
+
+  const defaultPackager = packagerFor(app);
+  app.package = function(inputTree) {
+    const analysedTree = new BroccoliConcatAnalyser(inputTree);
+
+    return defaultPackager(analysedTree);
+  }
+
+  return app.toTree();
+}
+```
+
+#### Static Assets Split
+
+One of the techniques for improving site speed is isolating changes throughout
+application deployments. Assuming  the application assets are uploaded to CDN,
+the reasoning is very simple: if `ember.js` or `jQuery` (possibly along with
+other third party libraries) don't change with every deployment, why bust CDN
+cache for them?
+
+#### ES6 Modules
+
+ES6 modules are [starting](https://caniuse.com/#feat=es6-module) to land in browsers.
+This means that you can use `<script type="module" src="/my/app.js"></script>`.
+
+This [article](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/) [explains](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/#is-this-really-worth-the-extra-effort) the benefits of using ES6 modules over ES2015 (smaller total file sizes, faster to parse and evaluate).
+
+`package` API will make it possible to package your application for both ES2015 only browsers as well
+the ones with ES6 modules support.
+
+# Topics for Future RFCs
+
+While working on this RFC, some ideas were brought into focus regarding existing
+and new features in Ember CLI. They all likely require separate discussions in
+future RFCs, but the discussion points have been included below.
+
+## Tree-shaking
+
+Firstly, what's _tree-shaking_? AFAIK, the term
+[originated](https://groups.google.com/forum/#!msg/comp.lang.lisp/6zpZsWFFW18/-z_8hHRAIf4J)
+in Lisp. The gist of the idea is "how about we start using _only_ the code that
+we actually need?"
+
+Secondly, how is it different from [Dead Code
```

Comment:
You can similarly explain this as dead code elimination is an imperfect outside-in approach, and tree-shaking is a (near) perfect inside-out approach.